mesonbuild/modules/wayland.py

```diff
@@ -0,0 +1,115 @@
+        # wayland-scanner, as it's more robust
+        scanner_dep  = interpreter.dependency_impl(
+            'wayland-scanner', 'wayland-scanner', {'required': True, 'native': True})
+        scanner_name = scanner_dep.variable_method([], {'pkgconfig': 'wayland_scanner'})
+        self.scanner = interpreter.find_program_impl(
+            [scanner_name], silent=False, for_machine=MachineChoice.BUILD)
+
+        # Older version of wayland-scanner take different arguments
+        self.old_scanner = version_compare(scanner_dep.held_object.version, '< 1.15')
+
+        # We need this to find the location where upstream protocols are installed
+        self.protocols_dep = interpreter.dependency_impl(
+            'wayland-protocols', 'wayland-protocols', {'required': True})
+
+    @stringArgs
+    @permittedKwargs(build.CustomTarget.known_kwargs | {'is_protocol'})
+    def scanner_target(self, state: 'ModuleState', args: T.Sequence[str], kwargs: T.Dict[str, T.Any]) -> ModuleReturnValue:
+        """Wrapper the wayland-scanner tool, and create custom targets.
+
+        This provides a wrapper around wayland-scanner, simplifying things
+        like where to find the installed protocols and setting up paths
+        correctly. It can also auto generate an output name based on teh
+        input name, if an output name isn't provided, which is fine for
+        generating C files, though it may not be for generating headers.
+        """
+        if len(args) != 2:
+            raise MesonException('scanner_target requires 1 arguments')
+
+        name = args[0]
+        if not isinstance(name, str):
+            raise MesonException('First argument must be a string')
+
+        valid_types = {'client-header', 'server-header', 'private-code', 'public-code'}
+        type_ = args[1]
+        if not isinstance(type_, str):
+            raise MesonException('Second argument must be a string')
+        if type_ not in valid_types:
+            raise MesonException('Second argument must be one of {}'.format(' '.join(sorted(valid_types))))
+
+        # map new argument to old arguments if necessary (so we only accept the
+        # newer versions), error if the wayland-scanner is too old for the
+        # requested argument
+        if self.old_scanner:
+            if type_ == 'private-code':
+                type_ = 'code'
+            elif type_ == 'public-code':
+                raise MesonException('public-code type requires wayland scanner >=1.15')
+
+        if isinstance(kwargs['input'], list):
+            if len(kwargs['input']) != 1:
+                raise MesonException('Only one input file allowed.')
+            input_ = kwargs['input'][0]
+        else:
+            input_ = kwargs['input']
+        if kwargs.pop('is_protocol', False):
+            p = self.protocols_dep.variable_method([], {'pkgconfig': 'pkgdatadir'})
+            input_ = (PurePath(p) / input_).as_posix()
```

Comment:
I find `is_protocol` wording confusing, because everything is about protocol here.

Is there a reason to make a shorthand for `wayland-protocols` specifically? Other protocol repositories exist too, and I'm ok with the shorthand for `wayland-protocols` (which might not always be required), but how does one use extensions from other projects?

E.g. Weston installs some XML files and announces the location via `libweston-MAJOR-protocols.pc` file `pkgdatadir`. There is also https://github.com/swaywm/wlr-protocols .